flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java

```diff
@@ -19,30 +19,175 @@
+                                i -> {
+                                    if (i % numRecordsPerSplit == stopAtRecord1
+                                            || i % numRecordsPerSplit == stopAtRecord2) {
+                                        beforeBarrier.get().await();
+                                        afterBarrier.get().await();
+                                    }
+                                    return i;
+                                });
+        stream.addSink(new DiscardingSink<>());
+        JobClient jobClient = env.executeAsync();
+
+        beforeBarrier.get().await();
+        assertSourceMetrics(stopAtRecord1 + 1, numRecordsPerSplit, env.getParallelism(), numSplits);
+        afterBarrier.get().await();
+
+        beforeBarrier.get().await();
+        assertSourceMetrics(stopAtRecord2 + 1, numRecordsPerSplit, env.getParallelism(), numSplits);
+        afterBarrier.get().await();
+
+        jobClient.getJobExecutionResult().get();
+    }
+
+    private void assertSourceMetrics(
+            long processedRecordsPerSubtask,
+            long numTotalPerSubtask,
+            int parallelism,
+            int numSplits) {
+        List<OperatorMetricGroup> groups =
+                miniClusterResource.getMetricReporter().findOperatorMetricGroups("TestingSource");
+        assertThat(groups, hasSize(parallelism));
+
+        int subtaskWithMetrics = 0;
+        for (OperatorMetricGroup group : groups) {
+            Map<String, Metric> metrics =
+                    miniClusterResource.getMetricReporter().getMetricsByGroup(group);
+            // there are only 2 splits assigned; so two groups will not update metrics
+            if (group.getIOMetricGroup().getNumRecordsInCounter().getCount() == 0) {
+                // assert that optional metrics are not initialized when no split assigned
+                assertThat(metrics.get(MetricNames.CURRENT_EMIT_EVENT_TIME_LAG), nullValue());
+                assertThat(metrics.get(MetricNames.WATERMARK_LAG), nullValue());
+                continue;
+            }
+            subtaskWithMetrics++;
+            // I/O metrics
+            assertThat(
+                    group.getIOMetricGroup().getNumRecordsInCounter(),
+                    isCounter(equalTo(processedRecordsPerSubtask)));
+            assertThat(
+                    group.getIOMetricGroup().getNumBytesInCounter(),
+                    isCounter(
+                            equalTo(
+                                    processedRecordsPerSubtask
+                                            * MockRecordEmitter.RECORD_SIZE_IN_BYTES)));
+            // MockRecordEmitter is just incrementing errors every even record
+            assertThat(
+                    metrics.get(MetricNames.NUM_RECORDS_IN_ERRORS),
+                    isCounter(equalTo(processedRecordsPerSubtask / 2)));
+            // Timestamp assigner subtracting EVENTTIME_LAG from wall clock, so expect that lag
```

Comment:
Comment incomplete?

Reply:
Probably just bad english; I'll try to fix.